File: Projects/FlutterPerfLab/lib/modules/maptest/map_test_view_model.dart
import 'dart:async';
import 'dart:isolate';
import 'package:flutter/foundation.dart';
import 'package:share_plus/share_plus.dart';

import 'package:flutterperflab/core/managers/performance_manager.dart';
import 'package:flutterperflab/core/managers/export_manager.dart';
import 'package:flutterperflab/models/map_point.dart';
import 'package:flutterperflab/core/generator/map_test_data_generator.dart';

// Framework-agnostic dependencies
import 'map_controller_adapter.dart';
import 'camera_target.dart';

/// Manages the benchmark state and execution strictly in Dart, relying on
/// the platform channel initialization payload.
class MapTestViewModel extends ChangeNotifier {
  // MARK: - Benchmark Constants
  static const double _zoomOutLevel = 13.0;
  static const double _zoomInLevel = 17.5;

  // Google Maps and Apple Maps calculate visible coordinate spans differently.
  // We define platform-specific baselines here to ensure equal rendering payloads.
  static const CameraTarget _androidInitialTarget = CameraTarget(
    latitude: 41.0082,
    longitude: 28.9784,
    zoom: 11.12, // Google Maps Scale
  );

  static const CameraTarget _iosInitialTarget = CameraTarget(
    latitude: 41.0082,
    longitude: 28.9784,
    zoom: 11.74, // Apple Maps Scale
  );

  // MARK: - Reactive State
  bool isRunning = false;
  String status = "Ready";
  bool isReportReady = false;
  List<MapPoint> points = [];

  // MARK: - Internal Dependencies
  IMapController? _mapController;
  final PerformanceManager _performance = PerformanceManager.shared;
  bool _isTaskCancelled = false;

  /// Returns the hardware-aligned starting coordinate based on the requested OS.
  CameraTarget getInitialTarget({required bool isIOS}) {
    return isIOS ? _iosInitialTarget : _androidInitialTarget;
  }

  void setMapController(IMapController controller) {
    _mapController = controller;
  }

  // MARK: - Execution Pipeline

  Future<void> startTest({required bool isIOS}) async {
    if (_mapController == null) {
      _updateState(status: "Error: Rendering Engine not initialized.");
      return;
    }

    stopTest(isFinished: false);
    _isTaskCancelled = false;
    _updateState(
      isRunning: true,
      status: "Initializing Geospatial Dataset...",
      isReportReady: false,
    );

    // Offloads dataset generation to an isolate to prevent main thread stalls
    // prior to telemetry initialization.
    final List<MapPoint> newPoints = await Isolate.run(
      () => MapTestDataGenerator.generatePoints(20),
    );

    points = newPoints;
    _updateState(status: "Benchmarking in Progress...");

    _performance.startMonitoring();
    _startPinTour(isIOS: isIOS);
  }

  /// Systematic multi-stage animation sequence.
  /// Translates spatial coordinates into engine-agnostic targets, delegating
  /// the interpolation entirely to the respective underlying platform views.
  Future<void> _startPinTour({required bool isIOS}) async {
    await Future.delayed(const Duration(milliseconds: 1000));

    for (int i = 0; i < points.length; i++) {
      if (_isTaskCancelled || !isRunning) break;
      final MapPoint point = points[i];

      // STAGE 1: Transition (Horizontal Panning)
      _updateState(status: "Target ${i + 1} / ${points.length} -> Panning ✈️");
      await _mapController?.animateCamera(
        CameraTarget(
          latitude: point.coordinate.latitude,
          longitude: point.coordinate.longitude,
          zoom: _zoomOutLevel,
        ),
      );

      await Future.delayed(const Duration(milliseconds: 1800));
      if (_isTaskCancelled || !isRunning) break;

      // STAGE 2: Inspection (Vertical Zooming)
      _updateState(status: "Target ${i + 1} -> Inspecting Detail 🔍");
      await _mapController?.animateCamera(
        CameraTarget(
          latitude: point.coordinate.latitude,
          longitude: point.coordinate.longitude,
          zoom: _zoomInLevel,
        ),
      );

      await Future.delayed(const Duration(milliseconds: 1800));
      if (_isTaskCancelled || !isRunning) break;

      // STAGE 3: Ascension (Zoom Out)
      if (i < points.length - 1) {
        _updateState(status: "Target ${i + 1} -> Ascending ⬆️");
        await _mapController?.animateCamera(
          CameraTarget(
            latitude: point.coordinate.latitude,
            longitude: point.coordinate.longitude,
            zoom: _zoomOutLevel,
          ),
        );
        await Future.delayed(const Duration(milliseconds: 1200));
      }
    }

    // FINAL STAGE: Global Context Review
    if (!_isTaskCancelled && isRunning) {
      _updateState(status: "Tour Finalized!");
      // Returns to the OS-specific starting position.
      await _mapController?.animateCamera(getInitialTarget(isIOS: isIOS));
      await Future.delayed(const Duration(milliseconds: 3000));
    }

    if (!_isTaskCancelled) stopTest(isFinished: true);
  }

  // MARK: - Termination & Telemetry Export

  void stopTest({required bool isFinished}) {
    _isTaskCancelled = true;
    if (!isRunning) return;

    _performance.stopMonitoring();
    _updateState(
      isRunning: false,
      status: isFinished ? "Completed ✅" : "Terminated 🛑",
      isReportReady: isFinished,
    );
  }

  Future<void> exportResults() async {
    final file = await ExportManager.shared.generateCSV(
      logs: _performance.currentLogs,
      testName: "Map_Optimized_Tour_Hybrid",
      customSummary: "Total_Points,${points.length}",
    );
    if (file != null) await Share.shareXFiles([XFile(file.path)]);
  }

  // MARK: - State Segregation

  /// Centralized state mutator to manage Rebuild triggers.
  void _updateState({bool? isRunning, String? status, bool? isReportReady}) {
    if (isRunning != null) this.isRunning = isRunning;
    if (status != null) this.status = status;
    if (isReportReady != null) this.isReportReady = isReportReady;
    notifyListeners();
  }
}

File: Projects/NativeKotlinPerfLab/app/src/main/java/com/tez/nativekotlinperflabapp/ui/features/maptest/MapTestViewModel.kt
package com.tez.nativekotlinperflabapp.ui.features.maptest

import android.content.Context
import android.content.Intent
import androidx.core.content.FileProvider
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.tez.nativekotlinperflabapp.core.generators.MapTestDataGenerator
import com.tez.nativekotlinperflabapp.core.managers.ExportManager
import com.tez.nativekotlinperflabapp.core.managers.PerformanceManager
import com.tez.nativekotlinperflabapp.models.MapPoint
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Orchestrator that automates complex camera maneuvers across a dataset
 * to measure rendering performance.
 */
class MapTestViewModel : ViewModel() {

    private val zoomOutLevel = 13f
    private val zoomInLevel = 17.5f
    private val defaultCenter = LatLng(41.0082, 28.9784)

    private val _cameraPosition = MutableStateFlow(
        CameraPosition.fromLatLngZoom(defaultCenter, 11.12f)
    )
    val cameraPosition: StateFlow<CameraPosition> = _cameraPosition.asStateFlow()

    private val _points = MutableStateFlow<List<MapPoint>>(emptyList())
    val points: StateFlow<List<MapPoint>> = _points.asStateFlow()

    private val _isRunning = MutableStateFlow(false)
    val isRunning: StateFlow<Boolean> = _isRunning.asStateFlow()

    private val _status = MutableStateFlow("Ready")
    val status: StateFlow<String> = _status.asStateFlow()

    private val _isReportReady = MutableStateFlow(false)
    val isReportReady: StateFlow<Boolean> = _isReportReady.asStateFlow()

    private var testJob: Job? = null
    private val performance = PerformanceManager

    /**
     * Initiates the automated geospatial sequence.
     */
    fun startTest(context: Context) {
        stopTest(isFinished = false)

        _isRunning.value = true
        _status.value = "Initializing Geospatial Dataset..."
        _isReportReady.value = false

        val appContext = context.applicationContext

        viewModelScope.launch(Dispatchers.Main) {

            // Generates data on a background thread to keep the main thread responsive.
            val newPoints = withContext(Dispatchers.Default) {
                MapTestDataGenerator.generatePoints(count = 20)
            }

            _points.value = newPoints
            _status.value = "Automated Tour in Progress..."

            performance.startMonitoring(appContext)
            startPinTour()
        }
    }

    /**
     * Executes a systematic multi-stage animation sequence.
     */
    private fun startPinTour() {
        testJob?.cancel()

        testJob = viewModelScope.launch(Dispatchers.Main) {
            delay(1000)

            for ((index, point) in _points.value.withIndex()) {
                if (!isActive || !_isRunning.value) break

                _status.value = "Target ${index + 1} / ${_points.value.size} -> Panning ✈️"

                _cameraPosition.value = CameraPosition.Builder()
                    .target(point.coordinate)
                    .zoom(zoomOutLevel)
                    .build()

                delay(1800)

                if (!isActive || !_isRunning.value) break

                _status.value = "Target ${index + 1} -> Inspecting Detail 🔍"
                _cameraPosition.value = CameraPosition.Builder()
                    .target(point.coordinate)
                    .zoom(zoomInLevel)
                    .build()

                delay(1800)

                if (!isActive || !_isRunning.value) break

                if (index < _points.value.size - 1) {
                    _status.value = "Target ${index + 1} -> Ascending ⬆️"
                    _cameraPosition.value = CameraPosition.Builder()
                        .target(point.coordinate)
                        .zoom(zoomOutLevel)
                        .build()

                    delay(1200)
                }
            }

            if (isActive && _isRunning.value) {
                _status.value = "Tour Finalized! Global Review 🌍"

                _cameraPosition.value = CameraPosition.Builder()
                    .target(defaultCenter)
                    .zoom(11.2f)
                    .build()

                delay(3000)
                stopTest(isFinished = true)
            }
        }
    }

    fun stopTest(isFinished: Boolean = false) {
        testJob?.cancel()
        testJob = null

        if (!_isRunning.value) return

        performance.stopMonitoring()
        _isRunning.value = false
        _status.value = if (isFinished) "Tour Completed ✅" else "Test Terminated 🛑"

        _isReportReady.value = isFinished
    }

    /**
     * Defers CSV generation and export to user intent.
     * Offloads disk I/O to a background thread to prevent UI freezing.
     */
    fun exportResults(context: Context) {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val summary = "Total_Points_Visited,${_points.value.size},,,,"
                val csvFile = ExportManager.generateCSV(
                    context = context,
                    logs = performance.currentLogs.value,
                    testName = "Map_Native_Optimized_Tour",
                    customSummary = summary
                )

                csvFile?.let { file ->
                    val uri = FileProvider.getUriForFile(
                        context,
                        "${context.packageName}.provider",
                        file
                    )

                    withContext(Dispatchers.Main) {
                        val shareIntent = Intent(Intent.ACTION_SEND).apply {
                            type = "text/csv"
                            putExtra(Intent.EXTRA_STREAM, uri)
                            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                        }
                        context.startActivity(Intent.createChooser(shareIntent, "Export Results (CSV)"))
                    }
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    _status.value = "❌ Export Failed: ${e.message}"
                }
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        if (_isRunning.value) performance.stopMonitoring()
        testJob?.cancel()
    }
}